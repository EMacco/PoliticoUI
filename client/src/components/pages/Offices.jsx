import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { toast } from 'react-toastify';
import AuthHeader from '../layouts/AuthHeader';
import AuthFooter from '../layouts/AuthFooter';
import AuthNav from '../layouts/AuthNav';
import Card from '../layouts/Card';
import { fetchGovernmentOffices, expressInterest } from '@actions/office';

class Offices extends Component {
  componentDidMount() {
    const {
      fetchGovernmentOffices,
      user: { id }
    } = this.props;
    fetchGovernmentOffices(id);
  }

  expressInterest = e => {
    const { expressInterest } = this.props;
    const { officeId, partyId } = JSON.parse(e.target.id);
    if (!partyId) return toast.error('Please join a political party before contesting');
    expressInterest(partyId, officeId);
  };

  render() {
    const { offices } = this.props;
    const {
      user: { partyid: partyId }
    } = this.props;

    let officesCard = [];
    offices.map((office, ind) => {
      const {
        officeId,
        officeLogo,
        officeName,
        officeType,
        electionDate,
        count,
        candidateStatus,
        officeBtn,
        alreadyApplied
      } = office;

      const officeContent = [
        {
          title: 'Type',
          value: officeType
        },
        {
          title: 'Date',
          value: electionDate
        },
        {
          officeId,
          officeName,
          numberOfCandidates: count,
          type: 'link'
        },
        {
          text: candidateStatus,
          style: officeBtn,
          type:
            candidateStatus === 'Approved candidate' || candidateStatus === 'Awaiting approval'
              ? 'span'
              : !alreadyApplied
              ? 'button'
              : 'empty',
          onClick: !alreadyApplied ? this.expressInterest : null,
          id: JSON.stringify({ officeId, partyId })
        }
      ];

      officesCard.push(
        <Fragment key={Math.random().toString()}>
          <Card name={officeName} image={officeLogo} content={officeContent} />
          {offices.length % 2 !== 0 && ind === offices.length - 1 ? <Card hidden="true" /> : ''}
        </Fragment>
      );
    });

    return (
      <div>
        <AuthHeader />
        <main id="user-home-main">
          <div className="container">
            <AuthNav current="offices" />
            <div id="loadingIconContainer" />
            <section id="user-home-content-div">
              <div className="profile-page-section">
                <h3 id="interested-office-header">Government Offices ({officesCard.length})</h3>
                <div className="content-group" id="governmentOfficeSlot">
                  {officesCard}
                </div>
              </div>
            </section>
          </div>
          <AuthFooter />
        </main>
      </div>
    );
  }
}

Offices.propTypes = {
  fetchGovernmentOffices: PropTypes.func.isRequired,
  offices: PropTypes.array.isRequired
};

const mapStateToProps = state => ({
  user: state.auth.user,
  offices: state.offices.offices
});

export default connect(
  mapStateToProps,
  { fetchGovernmentOffices, expressInterest }
)(Offices);
