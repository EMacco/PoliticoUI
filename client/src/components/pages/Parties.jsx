import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import AuthHeader from '../layouts/AuthHeader';
import AuthFooter from '../layouts/AuthFooter';
import AuthNav from '../layouts/AuthNav';
import Card from '../layouts/Card';
import { fetchPoliticalParties, changeUserParty } from '@actions/party';

class Parties extends Component {
  componentDidMount() {
    const { fetchPoliticalParties } = this.props;
    fetchPoliticalParties();
  }

  joinParty = e => {
    const { changeUserParty } = this.props;
    changeUserParty(e.target.id);
  };

  render() {
    const {
      parties,
      user: { partyid }
    } = this.props;

    let partiesCard = [];
    parties.map((party, ind) => {
      const { partyId, partyLogo, partyName, hqAddress } = party;

      const member = partyid === partyId ? true : false;
      let partyBtn = 'add-party-btn';
      let partyText = 'Join Party';
      // Check if this user is a member of this party
      if (partyid === partyId) {
        partyBtn = 'casted-vote';
        partyText = 'Member';
      }

      const partyContent = [
        {
          title: 'HQ Address',
          value: hqAddress
        },
        {
          text: partyText,
          style: partyBtn,
          type: member ? 'span' : 'button',
          onClick: !member ? this.joinParty : null,
          id: partyId
        }
      ];

      partiesCard.push(
        <Fragment key={Math.random().toString()}>
          <Card name={partyName} image={partyLogo} content={partyContent} />
          {parties.length % 2 !== 0 && ind === parties.length - 1 ? <Card hidden="true" /> : ''}
        </Fragment>
      );
    });

    return (
      <div>
        <AuthHeader />
        <main id="user-home-main">
          <div className="container">
            <AuthNav current="parties" />
            <div id="loadingIconContainer" />
            <section id="user-home-content-div">
              <div className="profile-page-section">
                <h3>Political Parties ({parties.length})</h3>
                <div className="content-group" id="userVotesSlot">
                  {partiesCard}
                </div>
              </div>
            </section>
          </div>
          <AuthFooter />
        </main>
      </div>
    );
  }
}

Parties.propTypes = {
  parties: PropTypes.array.isRequired,
  fetchPoliticalParties: PropTypes.func.isRequired,
  user: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  parties: state.parties.parties,
  user: state.auth.user
});

export default connect(
  mapStateToProps,
  { fetchPoliticalParties, changeUserParty }
)(Parties);
