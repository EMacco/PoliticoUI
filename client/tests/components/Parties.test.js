import React from 'react';
import { Parties } from '@components/pages/Parties';

const props = {
  fetchPoliticalParties: () => {},
  user: { partyid: 2 },
  parties: [
    {
      hqaddress: 'festac lagos',
      id: 1,
      logourl:
        'https://res.cloudinary.com/dn4pokov0/image/upload/v1551452104/dggfgdaqi5demtki1dnt.png',
      name: 'peoples democratic party',
      partyId: 1,
      partyLogo: '',
      partyName: 'Party Name'
    },
    {
      hqaddress: 'festac lagos',
      id: 2,
      logourl:
        'https://res.cloudinary.com/dn4pokov0/image/upload/v1551452220/ukhkvobjnabok4ymt7g3.jpg',
      name: 'all progressives congress',
      partyId: 2,
      partyLogo: '',
      partyName: 'Party Name'
    },
    {
      hqaddress: 'lekki, lagos',
      id: 3,
      logourl:
        'https://res.cloudinary.com/dn4pokov0/image/upload/v1551466591/aou6cohppt8yckombbsm.png',
      name: 'all progressives grand alliance',
      partyId: 3,
      partyLogo: '',
      partyName: 'Party Name'
    },
    {
      hqaddress: 'google, universal',
      id: 4,
      logourl:
        'https://res.cloudinary.com/dn4pokov0/image/upload/v1557735076/ogpffpwqi1s8hsuewlah.jpg',
      name: 'node package manager',
      partyId: 4,
      partyLogo: '',
      partyName: 'Party Name'
    }
  ]
};

const setup = () => {
  const wrapper = shallow(<Parties {...props} />);
  return wrapper;
};

describe('<Parties />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = setup();
  });

  it('should render the Parties component correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('Parties')).toBeTruthy();
  });
});
