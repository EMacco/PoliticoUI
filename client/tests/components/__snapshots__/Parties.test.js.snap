// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Parties /> should render the Parties component correctly 1`] = `
<div>
  <Connect(withRouter(AuthHeader)) />
  <main
    id="user-home-main"
  >
    <div
      className="container"
    >
      <Connect(withRouter(AuthNav))
        current="parties"
      />
      <div
        id="loadingIconContainer"
      />
      <section
        id="user-home-content-div"
      >
        <div
          className="profile-page-section"
        >
          <h3>
            Political Parties (
            4
            )
          </h3>
          <div
            className="content-group"
            id="userVotesSlot"
          >
            <Card
              content={
                Array [
                  Object {
                    "title": "HQ Address",
                    "value": undefined,
                  },
                  Object {
                    "id": 1,
                    "onClick": [Function],
                    "style": "add-party-btn",
                    "text": "Join Party",
                    "type": "button",
                  },
                ]
              }
              hidden="false"
              image=""
              name="Party Name"
            />
            <Card
              content={
                Array [
                  Object {
                    "title": "HQ Address",
                    "value": undefined,
                  },
                  Object {
                    "id": 2,
                    "onClick": null,
                    "style": "casted-vote",
                    "text": "Member",
                    "type": "span",
                  },
                ]
              }
              hidden="false"
              image=""
              name="Party Name"
            />
            <Card
              content={
                Array [
                  Object {
                    "title": "HQ Address",
                    "value": undefined,
                  },
                  Object {
                    "id": 3,
                    "onClick": [Function],
                    "style": "add-party-btn",
                    "text": "Join Party",
                    "type": "button",
                  },
                ]
              }
              hidden="false"
              image=""
              name="Party Name"
            />
            <Card
              content={
                Array [
                  Object {
                    "title": "HQ Address",
                    "value": undefined,
                  },
                  Object {
                    "id": 4,
                    "onClick": [Function],
                    "style": "add-party-btn",
                    "text": "Join Party",
                    "type": "button",
                  },
                ]
              }
              hidden="false"
              image=""
              name="Party Name"
            />
          </div>
        </div>
      </section>
    </div>
    <AuthFooter />
  </main>
</div>
`;
